File: Arrays/5_NextGreaterElementI.py
class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        d = {}
        st = []
        ans = []
        
        for x in nums2:
            while len(st) and st[-1] < x:
                d[st.pop()] = x
            st.append(x)

        for x in nums1:
            ans.append(d.get(x, -1))
            
        return ans

File: Arrays/5_NextGreaterElementI.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
    stack<int> s;
        unordered_map<int, int> m;
        for (int n : nums2) {
            while (s.size() && s.top() < n) {
                m[s.top()] = n;
                s.pop();
            }
            s.push(n);
        }
        vector<int> ans;
        for (int n : nums1) 
            ans.push_back(m.count(n) ? m[n] : -1);
        return ans;
    }
};
